src/jit/importer.cpp

```diff
@@ -18269,4 +18271,325 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
+    //   IL_021e:  callvirt   instance int32 System.Object::GetHashCode()
+    if ((objClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        JITDUMP("--- obj class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is a shared instance, in particular watch for System::__Canon.
+    // Perhaps overly broad...?
+    if ((objClassAttribs & CORINFO_FLG_SHAREDINST) != 0)
+    {
+        JITDUMP("--- obj class is shared instance class, sorry\n");
+        return;
+    }
+
+    // Fetch the method that would be called based on the declared type of 'this'
+    CORINFO_METHOD_HANDLE derivedMethod = info.compCompHnd->resolveVirtualMethod(baseMethod, objClass);
+
+    // If we failed to get a handle, we can't devirtualize.  This can
+    // happen when prejitting, if the devirtualization crosses
+    // servicing bubble boundaries.
+    if (derivedMethod == nullptr)
+    {
+        JITDUMP("--- no derived method, sorry\n");
+        return;
+    }
+
+    // Fetch attributes and do more vetting of the method. In general,
+    // if we can't inline, we won't devirtualize.
+    const DWORD derivedMethodAttribs = info.compCompHnd->getMethodAttribs(derivedMethod);
+    const bool  derivedMethodIsFinal = ((derivedMethodAttribs & CORINFO_FLG_FINAL) != 0);
+
+#if defined(DEBUG)
+    if (verbose)
+    {
+        const char* const derivedMethodNote = derivedMethodIsFinal ? " [final]" : "";
+        const char*       derivedMethodClassName;
+        const char* const derivedMethodName = eeGetMethodName(derivedMethod, &derivedMethodClassName);
+
+        printf("    devirt to %s::%s%s\n", derivedMethodClassName, derivedMethodName, derivedMethodNote);
+        gtDispTree(call);
+    }
+#endif // defined(DEBUG)
+
+    if (objClassIsFinal || derivedMethodIsFinal)
+    {
+        JITDUMP("!!! Inlining ok, no restrictions, final; can devirtualize\n");
+        const bool wasStubCall = call->IsVirtualStub();
+
+        // Make the updates.
+        call->gtFlags &= ~GTF_CALL_VIRT_VTABLE;
+        call->gtFlags &= ~GTF_CALL_VIRT_STUB;
+        call->gtCallMethHnd = derivedMethod;
+
+        // Virtual calls include an implicit null check, which we
+        // now need to make explicit. Not sure yet if we can
+        // restrict this to just a subset or need to do it for all,
+        // so will do it for all.
+        call->gtFlags |= GTF_CALL_NULLCHECK;
+
+        // Clear the inline candidate info (may be non-null since
+        // it's a union field used for other things by virtual
+        // stubs)
+        call->gtInlineCandidateInfo = nullptr;
+
+#ifdef FEATURE_READYTORUN_COMPILER
+        if (opts.IsReadyToRun())
+        {
+            // We should refetch the lookup value from the VM.
+            call->setEntryPoint(callInfo->codePointerLookup.constLookup);
+            call->gtCallMoreFlags &= ~GTF_CALL_M_VIRTSTUB_REL_INDIRECT;
+        }
+#endif
+
+        // Need to update call info too. This is fragile
```

Comment:
This is one area I'm still experimenting with.

I don't like the idea of patching the call info or the various context handles, and it seems like the patching may get more complex when extending this to support shared generics. 

But I am not sure we can anticipate all the info we might need early.

For instance, I'd like to call back into the devirtualization code during inlining or even later on during optimization, as we potentially learn more about types.